application.cpp

```diff
@@ -397,24 +412,58 @@ bool application::initialize_impl(int argc, char** argv, vector<abstract_plugin*
    return true;
 }
 
+void application::handle_exception(std::exception_ptr eptr, std::string_view origin) {
+   try {
+      if (eptr)
+         std::rethrow_exception(eptr);
+   } catch(const std::exception& e) {
+      std::cerr << "Caught " << origin << " exception: \"" << e.what() << "\"\n";
+   } catch(...) {
+      std::cerr << "Caught unknown " << origin << " exception.\n";
+   }
+}
+   
 void application::shutdown() {
+   std::exception_ptr eptr = nullptr;
+   
    for(auto ritr = running_plugins.rbegin();
        ritr != running_plugins.rend(); ++ritr) {
-      (*ritr)->shutdown();
+      try {
+         (*ritr)->shutdown();
+      } catch(...) {
+         eptr = std::current_exception();
+         handle_exception(eptr, (*ritr)->name());
+      }
    }
    for(auto ritr = running_plugins.rbegin();
        ritr != running_plugins.rend(); ++ritr) {
-      plugins.erase((*ritr)->name());
+      try {
+         plugins.erase((*ritr)->name());
+      } catch(...) {
+         eptr = std::current_exception();
+         std::string origin = (*ritr)->name() + " destructor";
+         handle_exception(eptr, origin);
+      }
+   }
+   try {
+      running_plugins.clear();
+      initialized_plugins.clear();
+      plugins.clear();
+   } catch(...) {
+      eptr = std::current_exception();
+      handle_exception(eptr, "plugin cleanup");
    }
-   running_plugins.clear();
-   initialized_plugins.clear();
-   plugins.clear();
    quit();
+
+   // if we caught an exception while shutting down a plugin, rethrow it so that main()
+   // can catch it and report the error
+   if (eptr)
+      std::rethrow_exception(eptr);
```

Comment:
What are your thoughts on throwing the first exception instead of the last?

Reply:
It is not quite clear to me what is best. First all exceptions are logged in the order they are received. And also (from what I see in leap) the plugins are initialized with the first one being the most critical, and shutdown in the reverse order. So if there was an exception in the first pluging to be shutdown (possibly the least critical), and another exception is the last pluging to be shutdown, maybe the last one would be the best one to rethrow.

Also, do we expect that the main program can do something meaningful based upon the exception received when shutting down a plugin? Maybe an all-or-nothing decision like whether to try restarting another instance of appbase or not?

In any case, I don't have a strong opinion one way or the other. I'm happy to change the code so that we rethrow the first excepttion received. If further exceptions are triggered by the first one we receive, it very well might be the right thing to do. I'm just not sure it is the case since the caught exceptions are when shutting down different plugins, but again I'm on the fence and happy to make the change.

Reply:
I think I would lean on throwing the first exception. @spoonincode what is your opinion?

Reply:
It's tricky in the current design to know what's best. But my best suggestion is that I feel like the first exception should be the one bubbled up because it was likely the crux of the failure.

Be aware that `exec()` is imo doing too much and causes another masking point here. If an exception is caught in `exec()` it's then rethrown _after_ `shutdown()`. That means if `shutdown()` decides to throw something too, the original exception bubbled to main is "overridden" by the shutdown exception. That could potentially mask the reason for the first exception -- even though `handle_exception()` would still log it, it is unable to print much utility from an `fc::exception` like when main could catch an `fc::exception`

Reply:
Thank you @heifner and @spoonincode .
I have modified the code in `shutdown()` to rethrow the first exception.
I have also modified the code in `exec()` so that if there was an exception in the mainloop, it is the one rethrown.
Does it seem OK now?